:construction: WIP : setup prometheus through Docker

replicajune opened the pull request:
this PR is related to #173 


there is some stuff left to do. Among others :

- setup docker installalation tasks in a dedicated install.yml file (as discussed in the issue)
- setup / adapt molecule to test installation with Docker

ATM, fluke8 seems to be broke. It runs now and as for tasks in blocks to be indent with 2 more spaces, but if I do so, I got this :

```
--> Executing Flake8 on files found in /home/vagrant/ansible-prometheus/molecule/alternative/tests/...
Traceback (most recent call last):
  File "/home/vagrant/ansible-prometheus/.tox/py27-ansible27/bin/flake8", line 6, in <module>
    from flake8.main.cli import main
  File "/home/vagrant/ansible-prometheus/.tox/py27-ansible27/local/lib/python2.7/site-packages/flake8/main/cli.py", line 2, in <module>
    from flake8.main import application
  File "/home/vagrant/ansible-prometheus/.tox/py27-ansible27/local/lib/python2.7/site-packages/flake8/main/application.py", line 14, in <module>
    from flake8.main import options
  File "/home/vagrant/ansible-prometheus/.tox/py27-ansible27/local/lib/python2.7/site-packages/flake8/main/options.py", line 4, in <module>
    from flake8.main import vcs
  File "/home/vagrant/ansible-prometheus/.tox/py27-ansible27/local/lib/python2.7/site-packages/flake8/main/vcs.py", line 4, in <module>
    from flake8.main import mercurial
  File "/home/vagrant/ansible-prometheus/.tox/py27-ansible27/local/lib/python2.7/site-packages/flake8/main/mercurial.py", line 7, in <module>
    import configparser
  File "/home/vagrant/ansible-prometheus/.tox/py27-ansible27/local/lib/python2.7/site-packages/configparser.py", line 12, in <module>
    from backports.configparser import (
  File "/home/vagrant/ansible-prometheus/.tox/py27-ansible27/local/lib/python2.7/site-packages/backports/configparser/__init__.py", line 139, in <module>
    from backports.configparser.helpers import OrderedDict as _default_dict
  File "/home/vagrant/ansible-prometheus/.tox/py27-ansible27/local/lib/python2.7/site-packages/backports/configparser/helpers.py", line 193, in <module>
    dict(__metaclass__=abc.ABCMeta),
  File "/home/vagrant/ansible-prometheus/.tox/py27-ansible27/lib/python2.7/abc.py", line 87, in __new__
    cls = super(ABCMeta, mcls).__new__(mcls, name, bases, namespace)
TypeError: type() argument 1 must be string, not newstr
An error occurred during the test sequence action: 'lint'. Cleaning up.
```
I'll do some comments inside the changes I made to explain a bit more clearly what I did here and there if there is any question about it

Also, I'm always looking for feedbacks so any comments are welcomed ! :smile:

paulfantom commented:
Why this is building prometheus image and not just use one from docker hub?

paulfantom commented on tasks/main.yml:
```diff
@@ -34,3 +34,19 @@
+    - prometheus_docker_install
+
+- name: test if prometheus is running
+  shell: sleep 2; curl -I -X GET localhost:9090/metrics 2> /dev/null |head -n1 |cut -d' ' -f2
```
This should be converted to use `uri` module. `shell` is considered an antipattern.

replicajune replied:
will do, thanks !

paulfantom commented on tasks/main.yml:
```diff
@@ -34,3 +34,19 @@
+
+- name: ensure container is up
+  become: true
+  shell: "{{ prometheus_docker_compose_exec.stdout }} --project-directory {{ prometheus_config_dir }}/docker.d -f {{ prometheus_config_dir }}/docker.d/docker-compose.yml up -d"
```
Ensuring that container is up should be done with `docker_container` module

replicajune replied:
ok, I wasn't aware of the docker modules in ansible. This make sense then that docker-compose isn't relevent when working with ansible !
:+1:

paulfantom commented on tasks/main.yml:
```diff
@@ -34,3 +34,19 @@
+  shell: sleep 2; curl -I -X GET localhost:9090/metrics 2> /dev/null |head -n1 |cut -d' ' -f2
+  register: prometheus_status_code
+
+- fail:
```
This wouldn't be needed if previous task was done with `uri` module and `retries` parameter.

replicajune replied:
:+1:

paulfantom commented on tasks/configure.yml:
```diff
@@ -57,7 +57,7 @@
     force: true
     owner: root
-    group: prometheus
+    group: "{% if prometheus_docker_install %} 65534 {% else %} prometheus {% endif %}"
```
Such operations should be set in preflight.yml and here it should be only a reference to some registered fact. Even better if those were set in some file in `vars/` directory.

replicajune replied:
got it ! :)

paulfantom commented on tasks/configure.yml:
```diff
@@ -4,9 +4,9 @@
+    group: "{% if prometheus_docker_install %} 65534 {% else %} prometheus {% endif %}"
     mode: 0640
-    validate: "/usr/local/bin/promtool check rules %s"
+    validate: "{{ prometheus_promtool_exec }} check rules %s"
```
`prometheus_promtool_exec` should be set in `vars/main.yml` and should be prefixed with `__`.

replicajune replied:
Got it !

just to be clear :
- I set this variables in vars with `/usr/local/bin/promtool` as value and override it if the prometheus_docker_install is triggered right  (in preflight) ?

Still i'm curious, what is the need to prefix this variable with __ ?

paulfantom replied:
`__` prefix is a common thing in python world to say that this variable is "private"

paulfantom commented on handlers/main.yml:
```diff
@@ -5,9 +5,25 @@
+
+- name: reload container
+  become: true
+  shell: "{{ prometheus_docker_compose_exec.stdout }} --project-directory {{ prometheus_config_dir }}/docker.d -f {{ prometheus_config_dir }}/docker.d/docker-compose.yml restart -d"
```
`shell` is an antipattern, this should be replaced with `docker_container`

paulfantom commented on handlers/main.yml:
```diff
@@ -5,9 +5,25 @@
+    - not prometheus_docker_install
+
+- name: restart container
+  become: true
```
This shouldn't be executed with `become: true` as it overwriting all security measures of target system

paulfantom commented on tasks/preflight.yml:
```diff
@@ -77,3 +79,64 @@
+    tags:
+      - skip_ansible_lint
+
+  - name: ensure that docker-compose is installed
```
there is no reason to use docker-compose.

paulfantom commented on tasks/preflight.yml:
```diff
@@ -77,3 +79,64 @@
+      - prometheus_docker_compose_extra_config != {}
+
+  - name: Get docker executable path
+    shell: whereis docker | awk '{print $2}'
```
If needed, there should be a check for `dockerd` (daemon) and not `docker` (client).

replicajune replied:
I should also propably use the stat module : https://docs.ansible.com/ansible/latest/modules/stat_module.html#stat-module

replicajune commented:
I've made a couple of changes regarding your remarks. In addition to that, I still need to :

- setup a proper separation between the installation methods (binaries / docker)
- manage molecule
- say a little something in the documentation on how to trigger an installation with docker :)

Regarding molecule, do you know if I can trigger a docker installation only on a specific platform  (in a molecule way) ?
From what I see, I would like to use booth existing setup (default and alternative) and trigger a docker installation over just one image (docker:dind). Any existing linuxes having docker should install prometheus with docker no matter if it's debian or centos so the only thing that need to be tested is docker I think .. if installing a docker image. Do you see what I mean ?

replicajune commented:
I'm trying to run already existing molecules tests but they fails. I have this (new) error :

```
    fatal: [stretch -> localhost]: FAILED! => {"attempts": 5, "changed": false, "msg": "Aborting, target uses selinux but python bindings (libselinux-python) aren't installed!"}                                                                                               
    FAILED - RETRYING: download prometheus binary to local folder (1 retries left).                                                                                                                                                                                             
    FAILED - RETRYING: download prometheus binary to local folder (5 retries left).                                                                                                                                                                                             
    fatal: [bionic -> localhost]: FAILED! => {"attempts": 5, "changed": false, "msg": "Aborting, target uses selinux but python bindings (libselinux-python) aren't installed!"}
    fatal: [jessie -> localhost]: FAILED! => {"attempts": 5, "changed": false, "msg": "Aborting, target uses selinux but python bindings (libselinux-python) aren't installed!"}
    fatal: [centos7 -> localhost]: FAILED! => {"attempts": 5, "changed": false, "msg": "Aborting, target uses selinux but python bindings (libselinux-python) aren't installed!"}
    FAILED - RETRYING: download prometheus binary to local folder (4 retries left).                                                     
    FAILED - RETRYING: download prometheus binary to local folder (3 retries left).                                                     
    fatal: [xenial -> localhost]: FAILED! => {"attempts": 5, "changed": false, "msg": "Aborting, target uses selinux but python bindings (libselinux-python) aren't installed!"}                                                                                                
    FAILED - RETRYING: download prometheus binary to local folder (2 retries left).                                                     
    FAILED - RETRYING: download prometheus binary to local folder (1 retries left).                                                                                                                                                                                             
    fatal: [fedora -> localhost]: FAILED! => {"attempts": 5, "changed": false, "msg": "Aborting, target uses selinux but python bindings (libselinux-python) aren't installed!"}                                                                                                
```

This happen at "create prometheus configuration directories" task

I'm probably missing something but I don't know what.

`libselinux-python` is installed on host but doesn't do much..

replicajune commented:
I'm trying to get on the molecule side of things to test an installation with docker. 

ATM, this is not working as I'm facing an issue with molecule (probably nothing really problematic but my limited knowlege with molecule blocks me for now)

The problem appear when molecule is in the converge action and complain that he can't get facts :

```
    TASK [Gathering Facts] *********************************************************
    fatal: [docker]: UNREACHABLE! => {"changed": false, "msg": "Authentication or permission failure. In some cases, you may have been able to authenticate and did not have permissions on the target directory. Consider changing the remote tmp path in ansible.cfg to a path rooted in \"/tmp\". Failed command was: ( umask 77 && mkdir -p \"` echo ~/.ansible/tmp/ansible-tmp-1549558006.05-243433001157362 `\" && echo ansible-tmp-1549558006.05-243433001157362=\"` echo ~/.ansible/tmp/ansible-tmp-1549558006.05-243433001157362 `\" ), exited with result 1", "unreachable": true}
``` 

This is because the container he tries to connect to has been destroy just before. Nothing on molecule tell me that though. Here's a more full trace of execution :

```
vagrant@ubuntu-xenial:~/ansible-prometheus$ molecule test -s docker
--> Validating schema /home/vagrant/ansible-prometheus/molecule/default/molecule.yml.
Validation completed successfully.
--> Test matrix
    
└── docker
    ├── lint
    ├── destroy
    ├── dependency
    ├── syntax
    ├── create
    ├── prepare
    ├── converge
    ├── idempotence
    ├── side_effect
    ├── verify
    └── destroy
    
--> Scenario: 'docker'
--> Action: 'lint'
--> Executing Yamllint on files found in /home/vagrant/ansible-prometheus/...
Lint completed successfully.
--> Executing Flake8 on files found in /home/vagrant/ansible-prometheus/molecule/default/tests/...
Lint completed successfully.
--> Executing Ansible Lint on /home/vagrant/ansible-prometheus/molecule/default/playbook.yml...
    
Lint completed successfully.
--> Scenario: 'docker'
--> Action: 'destroy'
    
    PLAY [Destroy] *****************************************************************
    
    TASK [Destroy molecule instance(s)] ********************************************
    changed: [localhost] => (item=None)
    changed: [localhost]
    
    TASK [Wait for instance(s) deletion to complete] *******************************
    ok: [localhost] => (item=None)
    ok: [localhost]
    
    TASK [Delete docker network(s)] ************************************************
    
    PLAY RECAP *********************************************************************
    localhost                  : ok=2    changed=1    unreachable=0    failed=0
    
    
--> Scenario: 'docker'
--> Action: 'dependency'
Skipping, missing the requirements file.
--> Scenario: 'docker'
--> Action: 'syntax'
    
    playbook: /home/vagrant/ansible-prometheus/molecule/default/playbook.yml
    
--> Scenario: 'docker'
--> Action: 'create'
    
    PLAY [Create] ******************************************************************
    
    TASK [Create molecule instance(s)] *********************************************
    changed: [localhost] => (item=None)
    changed: [localhost]
    
    TASK [Wait for instance(s) creation to complete] *******************************
    changed: [localhost] => (item=None)
    changed: [localhost]
    
    PLAY RECAP *********************************************************************
    localhost                  : ok=2    changed=2    unreachable=0    failed=0
    
    
--> Scenario: 'docker'
--> Action: 'prepare'
    
    PLAY [Prepare] *****************************************************************
    
    PLAY RECAP *********************************************************************
    
    
--> Scenario: 'docker'
--> Action: 'converge'
    
    PLAY [Run role] ****************************************************************
    
    TASK [Gathering Facts] *********************************************************
    fatal: [docker]: UNREACHABLE! => {"changed": false, "msg": "Authentication or permission failure. In some cases, you may have been able to authenticate and did not have permissions on the target directory. Consider changing the remote tmp path in ansible.cfg to a path rooted in \"/tmp\". Failed command was: ( umask 77 && mkdir -p \"` echo ~/.ansible/tmp/ansible-tmp-1549558006.05-243433001157362 `\" && echo ansible-tmp-1549558006.05-243433001157362=\"` echo ~/.ansible/tmp/ansible-tmp-1549558006.05-243433001157362 `\" ), exited with result 1", "unreachable": true}
    
    NO MORE HOSTS LEFT *************************************************************
    
    PLAY RECAP *********************************************************************
    docker                     : ok=0    changed=0    unreachable=1    failed=0
    
    
ERROR: 
An error occurred during the test sequence action: 'converge'. Cleaning up.
--> Scenario: 'docker'
--> Action: 'destroy'
    
    PLAY [Destroy] *****************************************************************
    
    TASK [Destroy molecule instance(s)] ********************************************
    changed: [localhost] => (item=None)
    changed: [localhost]
    
    TASK [Wait for instance(s) deletion to complete] *******************************
    changed: [localhost] => (item=None)
    changed: [localhost]
    
    TASK [Delete docker network(s)] ************************************************
    
    PLAY RECAP *********************************************************************
    localhost                  : ok=2    changed=2    unreachable=0    failed=0
```

When I look at the logs though, I see the container being ran (`shim containerd-shim started`) and destroyed (`shim reaped`). Then ansible try to get in the container but dockerd tells that the container is not running :

```
Feb 07 16:46:42 ubuntu-xenial containerd[12040]: time="2019-02-07T16:46:42.746708718Z" level=info msg="shim containerd-shim started" address="/containerd-shim/moby/7c1b8ce99befde09676ee468959cf1e3869e051ea7feadc5a94d1609f1917e44/shim.sock" debug=false pid=17335
Feb 07 16:46:42 ubuntu-xenial kernel: eth0: renamed from vethd6e1338
Feb 07 16:46:42 ubuntu-xenial kernel: IPv6: ADDRCONF(NETDEV_CHANGE): veth497c3b5: link becomes ready
Feb 07 16:46:42 ubuntu-xenial kernel: docker0: port 1(veth497c3b5) entered forwarding state
Feb 07 16:46:42 ubuntu-xenial kernel: docker0: port 1(veth497c3b5) entered forwarding state
Feb 07 16:46:43 ubuntu-xenial ansible-async_wrapper.py[17293]: Module complete (17293)
Feb 07 16:46:43 ubuntu-xenial containerd[12040]: time="2019-02-07T16:46:43.069769009Z" level=info msg="shim reaped" id=7c1b8ce99befde09676ee468959cf1e3869e051ea7feadc5a94d1609f1917e44
Feb 07 16:46:43 ubuntu-xenial dockerd[12182]: time="2019-02-07T16:46:43.080701027Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Feb 07 16:46:43 ubuntu-xenial kernel: docker0: port 1(veth497c3b5) entered disabled state
Feb 07 16:46:43 ubuntu-xenial kernel: vethd6e1338: renamed from eth0
Feb 07 16:46:43 ubuntu-xenial kernel: docker0: port 1(veth497c3b5) entered disabled state
Feb 07 16:46:43 ubuntu-xenial kernel: device veth497c3b5 left promiscuous mode
Feb 07 16:46:43 ubuntu-xenial kernel: docker0: port 1(veth497c3b5) entered disabled state
Feb 07 16:46:44 ubuntu-xenial ansible-async_wrapper.py[17218]: Done in kid B.
Feb 07 16:46:46 ubuntu-xenial dockerd[12182]: time="2019-02-07T16:46:46.075583751Z" level=error msg="Error setting up exec command in container docker: Container 7c1b8ce99befde09676ee468959cf1e3869e051ea7feadc5a94d1609f1917e44 is not running"
Feb 07 16:46:46 ubuntu-xenial dockerd[12182]: time="2019-02-07T16:46:46.109063857Z" level=error msg="Error setting up exec command in container docker: Container 7c1b8ce99befde09676ee468959cf1e3869e051ea7feadc5a94d1609f1917e44 is not running"
Feb 07 16:46:46 ubuntu-xenial dockerd[12182]: time="2019-02-07T16:46:46.143361032Z" level=error msg="Error setting up exec command in container docker: Container 7c1b8ce99befde09676ee468959cf1e3869e051ea7feadc5a94d1609f1917e44 is not running"
Feb 07 16:46:47 ubuntu-xenial ansible-async_wrapper.py[17583]: Invoked with 654432713094 7200 /home/vagrant/.ansible/tmp/ansible-tmp-1549558007.23-50339111557339/AnsiballZ_docker_container.py _
Feb 07 16:46:47 ubuntu-xenial ansible-async_wrapper.py[17583]: Return async_wrapper task started.
Feb 07 16:46:47 ubuntu-xenial ansible-async_wrapper.py[17586]: Starting module and watcher
Feb 07 16:46:47 ubuntu-xenial ansible-async_wrapper.py[17586]: Start watching 17587 (7200)
Feb 07 16:46:47 ubuntu-xenial ansible-async_wrapper.py[17587]: Start module (17587)
Feb 07 16:46:47 ubuntu-xenial ansible-async_wrapper.py[17292]: Done in kid B.
Feb 07 16:46:47 ubuntu-xenial ansible-async_wrapper.py[17587]: Module complete (17587)
Feb 07 16:46:52 ubuntu-xenial ansible-async_wrapper.py[17586]: Done in kid B.
```

And in fact when ansible try to get in the container `Container 7c1b8ce99befde09676ee468959cf1e3869e051ea7feadc5a94d1609f1917e44` just destroyed before, he fails.

replicajune commented:
I think my problem happening because I use docker:dind and the default behavior of this image is to be destroyed if nothing hang them up. I've dig up a bit in the other images used by molecule and they are all made to launch systemd. Once they're up they don't die and the converge step can go on

`docker:dind` uses alpine as a base so I  won't be able to use systemd. I'll try to find an image with supervisord or runit already built, if nothing came up, I'll setup one myself, and test my theory.

replicajune commented:
I've setup docker:dind to have supervisord as an entrypoint. It works but the role being compatible only with centos or debian, I have a fail at the very firest task of the role..

I'll have to find another base image :)

replicajune commented:
I've build localy an image using `billyteves/ubuntu-dind:16.04`. the issue I had is fixed. I'm still stuck on other tasks but I have made progress I hope to have something localy soon

replicajune commented:
I've built a docker-dind based on ubuntu with supervisord to mimic what's needed for molecule to test the role.

Its working up to the point where the container is suposed to be running. At that point, the container fails.
The thing is for the configuration file of prometheus, I'm doing a bind mount of the file to the container and it fails for molecule because all the initialisation of prometheus happen in a container. binds mounts though happens between the host and the prometheus container because the molecule container is built and build the prometheus container **with the same dockerd engine** that is **on the host**. Therefore when the container is starting, no config file is found :

```
    TASK [ansible-prometheus : ensure container is up] *****************************
    fatal: [docker]: FAILED! => {"changed": false, "msg": "Error starting container 8bfa98f0c0d57ea0afa6a84bcc3c24e353dcf83ce666be920385dd3029763296: 400 Client Error: Bad Request (\"{\"message\":\"OCI runtime create failed: container_linux.go:344: starting container process caused \\\"process_linux.go:424: container init caused \\\\\\\"rootfs_linux.go:58: mounting \\\\\\\\\\\\\\\"/srv/prometheus/config/prometheus.yml\\\\\\\\\\\\\\\" to rootfs \\\\\\\\\\\\\\\"/var/lib/docker/overlay2/efb5913f5cf582a7bb1e9ad101fe18660fb2340731a39c33471ab436a3a97db6/merged\\\\\\\\\\\\\\\" at \\\\\\\\\\\\\\\"/var/lib/docker/overlay2/efb5913f5cf582a7bb1e9ad101fe18660fb2340731a39c33471ab436a3a97db6/merged/etc/prometheus/prometheus.yml\\\\\\\\\\\\\\\" caused \\\\\\\\\\\\\\\"not a directory\\\\\\\\\\\\\\\"\\\\\\\"\\\": unknown: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type\"}\")"}
```

All other bind mounts will fail. The data volume was setup with a bind option to keep control of the location of prometheus data, but  if I let docker manage the volume, the container works with the volume, but the came can't be done with external files such as configuration files.

The role is working but I don't think that I can go much further on with molecule because of the way molecule and docker works with each other. A cleaner way would be to have a vagrant like driver for molecule to use instead of docker and work with that instead of docker containers.

Note : I've try to emphasis that he problem is happening with the docker engine, because of all the docker in docker happenning, things can be a bit confusing !

replicajune commented:
I've updated the documentation to explain usage of variables related to docker

paulfantom commented on .gitignore:
```diff
@@ -12,3 +12,5 @@
 __pycache__/
 .pytest_cache
 .tox
+Vagrantfile
```
not needed, we are not testing in vagrant :)

replicajune replied:
Exactly ! I was doing test localy on vagrant and kept the .gitignore with that for a while.. I guess I commited that by error :)

paulfantom commented on README.md:
```diff
@@ -16,11 +16,16 @@ Deploy [Prometheus](https://github.com/prometheus/prometheus) monitoring system
 
 When upgrading from <= 2.4.0 version of this role to >= 2.4.1 please turn off your prometheus instance. More in [2.4.1 release notes](https://github.com/cloudalchemy/ansible-prometheus/releases/tag/2.4.1)
 
+### Install with docker
```
For readability reasons this section should be mentioned after "Role variables".

paulfantom commented on README.md:
```diff
@@ -16,11 +16,16 @@ Deploy [Prometheus](https://github.com/prometheus/prometheus) monitoring system
 - Ansible >= 2.5 (It might work on previous versions, but we cannot guarantee it)
 - jmespath on deployer machine. If you are using Ansible from a Python virtualenv, install *jmespath* to the same virtualenv via pip.
 - gnu-tar on Mac deployer host (`brew install gnu-tar`)
+- If prometheus is installed with docker, docker must be installed prior to any deployment with this role, you cand use [this role](https://galaxy.ansible.com/geerlingguy/docker).
```
Please change role to https://github.com/mongrelion/ansible-role-docker. This is purely for marketing reasons :)

paulfantom commented on README.md:
```diff
@@ -45,6 +50,9 @@ All variables which can be overridden are stored in [defaults/main.yml](defaults
 | `prometheus_config_file` | "prometheus.yml.j2" | Variable used to provide custom prometheus configuration file in form of ansible template |
 | `prometheus_alert_rules` | [defaults/main.yml#L58](https://github.com/cloudalchemy/ansible-prometheus/blob/ff7830d06ba57be1177f2b6fca33a4dd2d97dc20/defaults/main.yml#L58) | Full list of alerting rules which will be copied to `{{ prometheus_config_dir }}/rules/ansible_managed.rules`. Alerting rules can be also provided by other files located in `{{ prometheus_config_dir }}/rules/` which have `*. ...
+| `prometheus_docker_install` | false | choose to install prometheus with docker |
+| `prometheus_docker_image` | quay.io/prometheus/prometheus |  docker image source of prometheus |
```
We kinda have a policy to disallow changes of source binary. In this case, variable should be defined in `vars/main.yml` or removed completely.

paulfantom commented on defaults/main.yml:
```diff
@@ -4,11 +4,27 @@ prometheus_version: 2.6.1
 prometheus_db_dir: /var/lib/prometheus
 
-prometheus_web_listen_address: "0.0.0.0:9090"
+prometheus_web_listen_port: 9090
```
Variable `prometheus_web_listen_address` cannot be removed as this is a variable defined by prometheus itself. 

I would advise to use that one and do `split(':')[-1]` to get port and `{{ var | split(':')[0:-1] | join(':')` to get address. Just remember that `:` char can be specified multiple times due to how IPv6 works.

replicajune replied:
Didn't knew that possibility. That said, I think it's more obvious to a reader to have variables instead of python function to build up variables (one for port, one for ip address, one for socket). I also didn't want to mess with variables already in place so I kinda merge my ideas in `prometheus_web_listen_address`. if I had to build this from scratch I would probably made the main variable be `prometheus_web_listen_socket` rather than ip..

paulfantom commented on defaults/main.yml:
```diff
@@ -4,11 +4,27 @@ prometheus_version: 2.6.1
 
+prometheus_docker_install: false
+
+prometheus_docker_image: "quay.io/prometheus/prometheus"
```
Variable should be either removed or placed in `vars/main.yml`

replicajune replied:
I agree with the policy because it makes things more clear as the official source is indeed only this one. I think having the possibility to overload that is convinient but I get your point too ! :)

paulfantom commented on defaults/main.yml:
```diff
@@ -4,11 +4,27 @@ prometheus_version: 2.6.1
+prometheus_docker_image: "quay.io/prometheus/prometheus"
+
+# # only strings allowed
+# prometheus_docker_labels:
```
Variable `prometheus_docker_labels` should be defined as an empty dict. The same as `prometheus_config_flags_extra` is.

paulfantom commented on handlers/main.yml:
```diff
@@ -5,9 +5,27 @@
+    - not prometheus_docker_install
+
+- name: reload prometheus container
+  shell: docker exec -t prometheus kill -HUP 1
```
Use `command`

paulfantom commented on molecule/docker/create.yml:
```diff
@@ -0,0 +1,37 @@
+---
```
Why this file is needed? Why not just link in `molecule.yml` to one from `defaults/` ?

replicajune replied:
regarding the molecule scenario, I basically copied the folder and tried to match the scenario to make molecule build prometheus with docker. I've then adapted things where needed but I've not cleaned anything nor did I had a deep enough understanding on how molecule was configured there. I was basicly modifying the scenario until I hit my dead end with it

paulfantom commented on molecule/docker/destroy.yml:
```diff
@@ -0,0 +1,32 @@
+---
```
Why this file is needed? Why not just link in `molecule.yml` to one from `defaults/` ?

paulfantom commented on molecule/docker/molecule.yml:
```diff
@@ -0,0 +1,33 @@
+  lint:
+    name: ansible-lint
+  playbooks:
+    create: create.yml
```
This should link to files from `default/`. Look at https://github.com/cloudalchemy/ansible-prometheus/blob/189ab57c67359cebfdbb4c951085f2c083d59f51/molecule/alternative/molecule.yml#L50-L53

replicajune replied:
I've used soft links on couple of repository when I needed the feature but not available with the tool I worked with, nice to see it's built in with molecule !

paulfantom commented on molecule/docker/molecule.yml:
```diff
@@ -0,0 +1,33 @@
+    volumes:
+      - /sys/fs/cgroup:/sys/fs/cgroup:ro
+      - /var/run/docker.sock:/var/run/docker.sock:rw
+      - /root/.ansible:/root/.ansible:rw
```
Why is anything from `/root` needed to be mounted to run test?

replicajune replied:
I've tried to avoid that a couple of times. I can't remeber if it was because of my setup (with vagrant) when molecule would be launch through a script from the Vagrantfile, being executed as root then.

paulfantom commented on molecule/docker/molecule.yml:
```diff
@@ -0,0 +1,33 @@
+      - /sys/fs/cgroup:/sys/fs/cgroup:ro
+      - /var/run/docker.sock:/var/run/docker.sock:rw
+      - /root/.ansible:/root/.ansible:rw
+      - /home:/home:rw
```
Why `/home` is mounted as read-write?

replicajune replied:
That is because I couldn't predict the user use to run ansible, is it molecule ?

paulfantom commented on molecule/docker/playbook.yml:
```diff
@@ -0,0 +1,11 @@
+    - ansible-prometheus
+  vars:
+    prometheus_docker_install: true
+    prometheus_version: latest
```
Please lock version and do not use `latest`. Using `latest` is causing flaky tests in CI due to GitHub rate limiter (`latest` needs to send a request to GitHub API)

paulfantom commented on molecule/docker/tests/test_default.py:
```diff
@@ -0,0 +1,69 @@
+        return yaml.load(stream)
+
+
+@pytest.mark.parametrize("dirs", [
```
This should match directories set in `molecule/docker/playbook.yml` with variables such as `prometheus_db_dir` and `prometheus_config_dir`

paulfantom commented on molecule/docker/tests/test_default.py:
```diff
@@ -0,0 +1,69 @@
+    assert d.exists
+
+
+@pytest.mark.parametrize("files", [
```
This should match directories set in `molecule/docker/playbook.yml` with variables such as `prometheus_db_dir` and `prometheus_config_dir`

paulfantom commented on molecule/docker/tests/test_default.py:
```diff
@@ -0,0 +1,69 @@
+    assert f.is_file
+
+
+@pytest.mark.parametrize("files", [
```
This should match directories set in `molecule/docker/playbook.yml` with variables such as `prometheus_db_dir` and `prometheus_config_dir`

paulfantom commented on molecule/docker/tests/test_default.py:
```diff
@@ -0,0 +1,69 @@
+    assert s.is_listening
+
+
+def test_version(host, AnsibleDefaults):
```
This won't work as prometheus binary is not in `/usr/local/bin/prometheus`

paulfantom commented on molecule/docker/tests/test_default.py:
```diff
@@ -0,0 +1,69 @@
+    assert not f.exists
+
+
+def test_service(host):
```
This won't work as there is no "service"

paulfantom commented on tasks/configure.yml:
```diff
@@ -4,39 +4,42 @@
     dest: "{{ prometheus_config_dir }}/rules/ansible_managed.rules"
     owner: root
-    group: prometheus
+    group: "{{ prometheus_id }}"
```
What is `prometheus_id` and how it is obtained. If this is something created in runtime then it should be prefixed with `__`.

replicajune replied:
it should be named prometheus_uid, and it's the uid of prometheus inside the container; should be the same inside and outside a container

paulfantom commented on tasks/install_docker.yml:
```diff
@@ -0,0 +1,16 @@
+---
+- name: Setup DB volume for prometheus
+  docker_volume:
+    name: prometheus_db
```
If volume is created for prometheus database then what is the behavior of `prometheus_db_dir`? Also why it is even changed in molecule test when it is obviously not used?

replicajune replied:
prometheus_db_dir was used to build the docker volume with a bind option to set the volume in the directory given by prometheus_db_dir, due to a dead end with molecule I change the behavior of that part (build the volume) and only build a 'basic volume' that was compatible with the way molecule test stuff.. but that move was more about making things works than anything as I couldn't do the same for config files in a simple/usable way

paulfantom commented on tasks/preflight.yml:
```diff
@@ -77,3 +79,39 @@
+        - prometheus_docker_labels != []
+
+    - name: Get Docker daemon executable path
+      shell: whereis dockerd | awk '{print $2}'
```
No `shell` please. This is an anti-pattern in ansible and should be used only as last resort.

paulfantom commented on tasks/preflight.yml:
```diff
@@ -77,3 +79,39 @@
+        - prometheus_docker_labels != []
+
+    - name: Get Docker daemon executable path
+      shell: whereis dockerd | awk '{print $2}'
```
Also you don't need a daemon, you need a client.

replicajune replied:
my guess is that at some point I would probably face a dead end because of all engine that would be able to run a container so.. the thing that should be check is the capacity to build a container rather than checking if a daemon or a client is there

paulfantom commented on tasks/preflight.yml:
```diff
@@ -77,3 +79,39 @@
+
+    - name: Override prometheus id
+      set_fact:
+        prometheus_id: 65534
```
Why an arbitrary id?

paulfantom commented on tasks/main.yml:
```diff
@@ -34,3 +34,32 @@
+    networks:
+      - name: prometheus
+    published_ports:
+      - "{{ prometheus_web_listen_port }}:{{ prometheus_web_listen_port }}"
```
What if I set following:
```
prometheus_web_listen_port: 10000
prometheus_web_listen_address: 0.0.0.0:5000
```
?

replicajune replied:
prometheus_web_listen_address is missleading I think but it that case it would be 10000. the variable `prometheus_web_listen_port` makes things more obvious about its usage where prometheus_web_listen_address does not only provide an address but a whole netsocket

paulfantom commented:
> The role is working but I don't think that I can go much further on with molecule because of the way molecule and docker works with each other. A cleaner way would be to have a vagrant like driver for molecule to use instead of docker and work with that instead of docker containers.

> Note : I've try to emphasis that he problem is happening with the docker engine, because of all the docker in docker happenning, things can be a bit confusing !

Since vagrant cannot be run (without hacks) in travis CI, docker is the only option. Additionally if this cannot be tested in CI, then I cannot merge it as this project is just an after-hours one and I don't want to spend a lot of time chasing bugs.

replicajune commented:
Hi,

I've anwser some of your feedbacks, and read all of them carefully :)

Thank you so much for all the feedback given over this PR, it was a first on a lot of levels, for me and my team, we've appreciated every feedback !!

Unfortunatly as you noticed I can't provide a better solution regarding the test made to build the container. I agree with you that we can't proceed further on this PR as this is holding any QA over this role.

I think that we're stuck until travis support vagrant (or any similar alternative) and molecule support any vm providers (I did not saw moelcule having a libvirt or virtualbox provider as well).

So until that is an option, or someone is having a better solution / conception of this idea, I think we can close this PR.

paulfantom commented:
I think you are right. Thanks for all the effort you put into creating this solution!
